app/lib/git_repository.rb

```diff
@@ -469,15 +475,38 @@ def add_directory(path)
   end
 
   # Removes a file from the repository.
-  def remove_file(path, expected_revision_identifier)
+  def remove_file(path, expected_revision_identifier, keep_folder: true)
     if @repos.last_commit.oid != expected_revision_identifier
       raise Repository::FileOutOfSyncConflict.new(path)
     end
     unless get_latest_revision.path_exists?(path)
       raise Repository::FileDoesNotExist.new(path)
     end
+    absolute_path = Pathname.new(File.join(@repos_path, path))
+    relative_path = Pathname.new(path)
     File.unlink(File.join(@repos_path, path))
     @repos.index.remove(path)
+    if keep_folder
+      unless File.exist?(File.join(absolute_path.dirname, DUMMY_FILE_NAME))
+        gitkeep_filename = File.join(relative_path.dirname, DUMMY_FILE_NAME)
+        add_file(gitkeep_filename, '')
+      end
+    end
+  end
+
+  def remove_directory(path, expected_revision_identifier)
+    if @repos.last_commit.oid != expected_revision_identifier
+      raise Repository::FileOutOfSyncConflict.new(path)
+    end
+    unless get_latest_revision.path_exists?(path)
+      raise Repository::FileDoesNotExist.new(path)
+    end
+    absolute_path = Pathname.new(File.join(@repos_path, path))
+    relative_path = Pathname.new(path)
+    FileUtils.remove_dir(absolute_path)
```

Comment:
This is better but I still think we should raise a conflict error if this directory is not empty and we're trying to delete it.